src/Bicep.Core/TypeSystem/Providers/Az/AzResourceTypeFactory.cs

```diff
@@ -127,60 +131,101 @@ private TypeSymbol ToTypeSymbol(Azure.Bicep.Types.Concrete.TypeBase typeBase, bo
+            bool isResourceBodyTopLevelPropertyType)
         {
-            if (!(extendedType.Type is Azure.Bicep.Types.Concrete.ObjectType objectType))
+            if (extendedType.Type is not Azure.Bicep.Types.Concrete.ObjectType objectType)
             {
                 throw new ArgumentException();
             }
 
-            var additionalProperties = objectType.AdditionalProperties != null ? GetTypeReference(objectType.AdditionalProperties) : null;
+            var additionalProperties = objectType.AdditionalProperties != null
+                ? GetTypeReference(objectType.AdditionalProperties, isResourceBodyType: false, isResourceBodyTopLevelPropertyType: isResourceBodyType)
+                : null;
 
             var extendedProperties = objectType.Properties.ToDictionary(kvp => kvp.Key, kvp => kvp.Value, StringComparer.OrdinalIgnoreCase);
             foreach (var property in baseProperties.Where(x => !extendedProperties.ContainsKey(x.Key)))
             {
                 extendedProperties[property.Key] = property.Value;
             }
 
-            return new ObjectType(name, GetValidationFlags(isResourceBodyType), extendedProperties.Select(kvp => GetTypeProperty(kvp.Key, kvp.Value)), additionalProperties, TypePropertyFlags.None);
+            return new ObjectType(name,
+                GetValidationFlags(isResourceBodyType, isResourceBodyTopLevelPropertyType),
+                extendedProperties.Select(kvp => GetTypeProperty(kvp.Key, kvp.Value, isResourceBodyType)),
+                additionalProperties,
+                TypePropertyFlags.None);
         }
 
-        private static TypeSymbolValidationFlags GetValidationFlags(bool isResourceBodyType)
+        private static TypeSymbolValidationFlags GetValidationFlags(bool isResourceBodyType, bool isResourceBodyTopLevelPropertyType)
         {
-            if (isResourceBodyType)
+            var flags = TypeSymbolValidationFlags.Default;
+
+            if (!isResourceBodyType)
             {
-                // strict validation on top-level resource properties, as 'custom' top-level properties are not supported by the platform
-                return TypeSymbolValidationFlags.Default;
+                flags |= TypeSymbolValidationFlags.WarnOnPropertyTypeMismatch;
             }
 
-            // in all other places, we should allow some wiggle room so that we don't block compilation if there are any swagger inaccuracies
-            return TypeSymbolValidationFlags.WarnOnTypeMismatch;
+            // strict validation on top-level resource properties, as 'custom' top-level properties are not supported by the platform
+            if (!isResourceBodyType && !isResourceBodyTopLevelPropertyType)
+            {
+                // in all other places, we should allow some wiggle room so that we don't block compilation if there are any swagger inaccuracies
+                flags |= TypeSymbolValidationFlags.WarnOnTypeMismatch;
+            }
```

Comment:
Conceptually, isn't the goal the same (error on top-level properties, warn on everything else)? Why do you need the new flag here?

Reply:
The flags are similar but handle slightly different cases. Consider the `properties` property on any resource: we know that this must be an object, and since it's part of the ARM RPC contract, we want to emit an error diagnostic if a template supplies a string or array instead of an object, but warn if there's a mismatch within the `properties` property

```bicep
resource res 'My.RP/widgets@2024-12-31' = {
  ...
  properties: 'foo' // <-- should be an error
}

resource res 'My.RP/widgets@2024-12-31' = {
  ...
  properties: {
    stringWithMinLength3: 'ab' // <-- should be a warning (string is too short)
    unknownProperty: 'foo' // <-- should be a warning (unexpected property)
  }
}
```

We're currently figuring out whether to warn or error for a property [based on the `WarnOnTypeMismatch` flag on the enclosing object](https://github.com/Azure/bicep/blob/091802dcc970c1a67898d3f59909292f29e5e134/src/Bicep.Core/TypeSystem/TypeValidator.cs#L1129). So `properties: 'foo'` above is an error because the body type for "My.RP/widgets@2024-12-31" resources does not have the `WarnOnTypeMismatch` flag, while the issues within `properties` are warnings because the `properties` property of the resource body type *does* have that flag. This presented an issue when I wanted to reuse a type that should generate a warning on mismatch (`resource<'My.RP/widgets@2024-12-31'>.properties.knownPropertyOfTypeString`) inside of an object that will not have the `WarnOnTypeMismatch` flag:

```bicep
// mod.bicep
param p resource<'My.RP/widgets@2024-12-31'>.properties.stringWithMinLength3

// main.bicep
module mod 'mod.bicep' = {
  name: 'mod'
  params: {
    p: 'ab'
  }
}
```

Since we don't want to set the `WarnOnTypeMismatch` flag on the `params` property of a module, I needed to update the property validator to look at a flag on the property type rather than on the enclosing object. Because that wouldn't cover unexpected properties, I added a secondary `WarnOnPropertyTypeMismatch` flag. For the "My.RP/widgets@2024-12-31" resource, the body type gets no flags, each top level property gets a `WarnOnPropertyTypeMismatch` flag, and everything else gets both `WarnOnPropertyTypeMismatch` and `WarnOnTypeMismatch`.